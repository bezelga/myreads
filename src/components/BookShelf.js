import React, {Component} from 'react'
import PropTypes from 'prop-types';
import Book from './Book'
import { shelvesDictionary } from '../utils'

class BookShelf extends Component {
  render() {
    return (
      <div className="bookshelf">
        <h2 className="bookshelf-title">{shelvesDictionary[this.props.status]}</h2>
        <div className="bookshelf-books">
          <ol className="books-grid">
            {this.props.books && this.props.books.map((book) => (
              <li key={book.title}>
                <Book
                  title={book.title}
                  authors={book.authors.join(', ')}
                  shelf={this.props.status}
                  coverURL={book.imageLinks && book.imageLinks.thumbnail}
                  changeBookStateFunction={this.props.changeBookStateFunction}
                />
              </li>
            ))}
          </ol>
        </div>
      </div>
    )
  }
}

BookShelf.propTypes = {
  books: PropTypes.array.isRequired
}

export default BookShelf
