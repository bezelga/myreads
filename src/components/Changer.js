import React, { Component } from 'react'
import PropTypes from 'prop-types';
import { shelvesDictionary } from '../utils'

class Changer extends Component {
  render () {
    return (
      <div className="book-shelf-changer">
        <select
         defaultValue={this.props.currentBookshelf}
         onChange={(event) => this.props.changeBookStateFunction(event.target.value)}
       >
          <option value="move" disabled>Move to...</option>
          {Object.entries(shelvesDictionary).map(([key, value]) => (
            <option key={key} value={key}>{value}</option>
          ))}
        </select>
      </div>
    )
  }
}

Changer.propTypes = {
  currentBookshelf: PropTypes.string.isRequired,
  changeBookStateFunction: PropTypes.func.isRequired
}

export default Changer
