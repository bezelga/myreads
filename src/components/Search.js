import React, { Component } from 'react'
import * as BooksAPI from '../BooksAPI'
import Book from './Book'
import { Link } from 'react-router-dom'
import debounce from 'lodash.debounce'

class Search extends Component {
  state = {
    query: '',
    books: null
  }


  cleanResults = () => {
    this.setState({ books: [] })
  }

  booksAPISearch = debounce((query) => {
    console.log('Debounced Event:', query);
    BooksAPI.search(query.trim()).then(this.afterSearchResponse)
  }, 500)

  updateQuery = (query) => {
    this.setState({ query: query, books: 'loading' })
    query ? this.booksAPISearch(query) : this.cleanResults()
  }

  afterSearchResponse = (books) => {
    if (Array.isArray(books)) {
      this.setState({ books })
    } else {
      this.setState({ books: [] })
    }
  }

  renderBook = (book) => {
    const shelf = this.props.booksMap[book.id] || "none"
    return (
      <li key={book.id}>
        <Book
          title={book.title}
          id={book.id}
          authors={book.authors}
          shelf={shelf}
          coverURL={book.imageLinks && book.imageLinks.thumbnail}
          changeBookStateFunction={this.props.changeBookStateFunction(shelf)}
        />
      </li>
    )
  }

  render() {
    const { query, books } = this.state

    return (
      <div className="search-books">
        <div className="search-books-bar">
          <Link to="/" className="close-search">Close</Link>
          <div className="search-books-input-wrapper">
            {/*
              NOTES: The search from BooksAPI is limited to a particular set of search terms.
              You can find these search terms here:
              https://github.com/udacity/reactnd-project-myreads-starter/blob/master/SEARCH_TERMS.md

              However, remember that the BooksAPI.search method DOES search by title or author. So, don't worry if
              you don't find a specific author or title. Every search is limited by search terms.
            */}
            <input
              value={query}
              onChange={(event) => this.updateQuery(event.target.value)}
              type="text"
              placeholder="Search by title or author"
            />
          </div>
        </div>
        <div className="search-books-results">
          <ol className="books-grid">
            {books === null ? (
              <div>Books will appear here</div>
            ) : (books === 'loading') ? (
              <div>Loading...</div>
            ) : (books.length === 0) ? (
              <div>No books found</div>
            ) : (books.map(book => (this.renderBook(book) )))}
          </ol>
        </div>
      </div>
    )
  }
}

export default Search
