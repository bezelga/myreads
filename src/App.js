import React from 'react'
import * as BooksAPI from './BooksAPI'
import './App.css'
import BookShelf from './components/BookShelf'
import Search from './components/Search'
import { Route, Link } from 'react-router-dom'

class BooksApp extends React.Component {
  state = {
    currentlyReading: [],
    wantToRead: [],
    read: [],
    none: []
  }

  componentDidMount() {
    BooksAPI.getAll().then((books) => {
      this.setState({
        currentlyReading: this.booksByShelf(books, "currentlyReading"),
        wantToRead: this.booksByShelf(books, "wantToRead"),
        read: this.booksByShelf(books, "read")
      })
    })
  }

  booksMap = () => {
    const result = {}

    Object.entries(this.state).forEach(([shelf, books]) => {
      books.forEach(book => (result[book.id] = shelf))
    })

    return result
  }

  booksByShelf = (books, shelf) => (
    books.filter(book => (book.shelf === shelf))
  )

  updateBookShelf = (currentShelf, newShelf) => (book) => {
    BooksAPI.update(book, newShelf)

    this.setState({
      [currentShelf]: this.state[currentShelf].filter((currentShelfBook) => (
        currentShelfBook.id !== book.id
      )),
      [newShelf]: [...this.state[newShelf], book]
    })
  }

  findBook = (shelf, bookID) => {
    const book = this.state[shelf].find(book => ( book.id === bookID ))
    return (book ? Promise.resolve(book) : BooksAPI.get(bookID))
  }

  changeBookState = (currentShelf) => (bookID) => (newShelf) => {
    if (currentShelf !== newShelf) {
      this.findBook(currentShelf, bookID).then(this.updateBookShelf(currentShelf, newShelf))
    }
  }

  render() {
    return (
      <div className="app">
        <Route exact path="/" render={() => (
          <div className="list-books">
            <div className="list-books-title">
              <h1>MyReads</h1>
            </div>
            <div className="list-books-content">
              <div>
                <BookShelf
                  status="currentlyReading"
                  books={this.state.currentlyReading}
                  changeBookStateFunction={this.changeBookState("currentlyReading")}
                />
                <BookShelf
                  status="wantToRead"
                  books={this.state.wantToRead}
                  changeBookStateFunction={this.changeBookState("wantToRead")}
                />
                <BookShelf
                  status="read"
                  books={this.state.read}
                  changeBookStateFunction={this.changeBookState("read")}
                />
              </div>
            </div>
            <div className="open-search">
              <Link to="/search">Add a book</Link>
            </div>
          </div>
        )}/>
        <Route exact path="/search" render={() => (
          <Search booksMap={this.booksMap()} changeBookStateFunction={this.changeBookState}/>
        )}/>
      </div>
    )
  }
}

export default BooksApp
