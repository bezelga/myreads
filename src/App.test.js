import React from 'react'
import ReactDOM from 'react-dom'
import App from './App'
import BookShelf from './components/BookShelf'
import Changer from './components/Changer'
import Search from './components/Search'
import { MemoryRouter } from 'react-router-dom'
import renderer from 'react-test-renderer'

/**
 This course is not designed to teach Test Driven Development.
 Feel free to use this file to test your application, but it
 is not required.
**/

jest.mock('./BooksAPI')

it('renders without crashing', () => {
  const div = document.createElement('div')
  ReactDOM.render(<MemoryRouter><App /></MemoryRouter>, div)
})

it('matches the contract with final snapshot', () => {
  const component = renderer.create(
    <MemoryRouter><App /></MemoryRouter>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
})

const changeBookStateFunction = () => () => {}

it('matches the bookshelf contract with final snapshot', () => {
  const component = renderer.create(
    <BookShelf changeBookStateFunction={changeBookStateFunction} books={[]}/>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
})

it('matches the changer contract with final snapshot', () => {
  const component = renderer.create(
    <Changer currentBookshelf="read" changeBookStateFunction={changeBookStateFunction}/>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
})

it('matches the search contract with final snapshot', () => {
  const component = renderer.create(
    <MemoryRouter><Search booksMap={{}} changeBookStateFunction={changeBookStateFunction}/></MemoryRouter>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
})
