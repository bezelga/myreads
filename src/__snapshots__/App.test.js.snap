// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the bookshelf contract with final snapshot 1`] = `
<div
  className="bookshelf"
>
  <h2
    className="bookshelf-title"
  />
  <div
    className="bookshelf-books"
  >
    <ol
      className="books-grid"
    />
  </div>
</div>
`;

exports[`matches the changer contract with final snapshot 1`] = `
<div
  className="book-shelf-changer"
>
  <select
    defaultValue="read"
    onChange={[Function]}
  >
    <option
      disabled={true}
      value="move"
    >
      Move to...
    </option>
    <option
      value="currentlyReading"
    >
      Currently Reading
    </option>
    <option
      value="wantToRead"
    >
      Want to Read
    </option>
    <option
      value="read"
    >
      Read
    </option>
    <option
      value="none"
    >
      None
    </option>
  </select>
</div>
`;

exports[`matches the contract with final snapshot 1`] = `
<div
  className="app"
>
  <div
    className="list-books"
  >
    <div
      className="list-books-title"
    >
      <h1>
        MyReads
      </h1>
    </div>
    <div
      className="list-books-content"
    >
      <div>
        <div
          className="bookshelf"
        >
          <h2
            className="bookshelf-title"
          >
            Currently Reading
          </h2>
          <div
            className="bookshelf-books"
          >
            <ol
              className="books-grid"
            />
          </div>
        </div>
        <div
          className="bookshelf"
        >
          <h2
            className="bookshelf-title"
          >
            Want to Read
          </h2>
          <div
            className="bookshelf-books"
          >
            <ol
              className="books-grid"
            />
          </div>
        </div>
        <div
          className="bookshelf"
        >
          <h2
            className="bookshelf-title"
          >
            Read
          </h2>
          <div
            className="bookshelf-books"
          >
            <ol
              className="books-grid"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="open-search"
    >
      <a
        href="/search"
        onClick={[Function]}
      >
        Add a book
      </a>
    </div>
  </div>
</div>
`;

exports[`matches the search contract with final snapshot 1`] = `
<div
  className="search-books"
>
  <div
    className="search-books-bar"
  >
    <a
      className="close-search"
      href="/"
      onClick={[Function]}
    >
      Close
    </a>
    <div
      className="search-books-input-wrapper"
    >
      <input
        onChange={[Function]}
        placeholder="Search by title or author"
        type="text"
        value=""
      />
    </div>
  </div>
  <div
    className="search-books-results"
  >
    <ol
      className="books-grid"
    >
      <div>
        Books will appear here
      </div>
    </ol>
  </div>
</div>
`;
